kong/runloop/wasm.lua

```diff
@@ -0,0 +1,646 @@
+
+    local routes = state.combined[service_id]
+
+    if not routes then
+      routes = {}
+      state.combined[service_id] = routes
+    end
+
+    routes[route_id] = ref
+
+  else
+    -- unreachable
+    error("unknown filter chain type: " .. tostring(typ), 2)
+  end
+end
+
+
+---
+-- Create a log-friendly string label for a filter chain reference.
+--
+---@param service_id? string
+---@param route_id?   string
+---@return string label
+local function label_for(service_id, route_id)
+  if service_id and route_id then
+    return "combined " ..
+           "service(" .. service_id .. "), " ..
+           "route(" .. route_id .. ")"
+
+  elseif service_id then
+    return "service(" .. service_id .. ")"
+
+  elseif route_id then
+    return "route(" .. route_id .. ")"
+
+  else
+    -- unreachable
+    error("can't compute a label for a filter chain with no route/service", 2)
+  end
+end
+
+
+---
+-- Build a combined filter list from 1-2 filter chain entities.
+--
+-- Disabled filter chains are skipped, and disabled filters are
+-- skipped.
+--
+-- Returns `nil` if no enabled filters are found.
+--
+---@param service_chain? kong.db.schema.entities.filter_chain
+---@param route_chain?   kong.db.schema.entities.filter_chain
+---
+---@return kong.db.schema.entities.wasm_filter[]?
+local function build_filter_list(service_chain, route_chain)
+  ---@type kong.db.schema.entities.wasm_filter[]|nil
+  local combined
+  local n = 0
+
+  if service_chain and service_chain.enabled then
+    for _, filter in ipairs(service_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  if route_chain and route_chain.enabled then
+    for _, filter in ipairs(route_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  return combined
+end
+
+
+---
+-- Unconditionally rebuild and return a new wasm state table from the db.
+--
+---@param  db                       table # kong.db
+---@param  version                  any
+---@param  old_state                kong.runloop.wasm.state
+---@return kong.runloop.wasm.state? new_state
+---@return string?                  err
+local function rebuild_state(db, version, old_state)
+  ---@type kong.db.schema.entities.filter_chain[]
+  local route_chains = {}
+
+  ---@type table<string, kong.db.schema.entities.filter_chain>
+  local service_chains_by_id = {}
+
+  ---@type kong.runloop.wasm.state
+  local state = {
+    by_service = {},
+    by_route = {},
+    combined = {},
+    version = version,
+  }
+
+  ---@type kong.runloop.wasm.filter_chain_reference[]
+  local all_chain_refs = {}
+
+  local page_size = db.filter_chains.max_page_size
+
+  for chain, err in db.filter_chains:each(page_size) do
+    if err then
+      return nil, "failed iterating filter chains: " .. tostring(err)
```

Comment:
```suggestion
      return nil, "failed iterating wasm filter chains: " .. tostring(err)
```

I'm curious how consistent the error logs pertaining to wasm really are. I think I have seen several permutations so far:
- Wasm
- wasm
- wasm filter chains
- Wasm filter chains

With or without quotes sometimes, or sometimes "wasm" in the beginning, the middle or the end of the error. Perhaps it isn't too bad, I haven't run into all of these errors myself during usage to see a busy logs file.

In ngx_wasm_module we are very strict about the format of each error to maintain high UX consistency.

Reply:
This is a good point to raise. The error strings in the branch have evolved over time with several different authors, so... entropy. I'll do another pass over the code and make things more consistent.

For errors that are logged within `kong.runloop.wasm` I have specifically avoided adding `wasm` to the message because we are already getting this context from `ngx.log()`.